frontend/packages/aquascope-editor/src/editor-utils/stepper.tsx

```diff
@@ -0,0 +1,383 @@
+    // Only one table widget can (should) be on a line.
+    return this.line.number === other.line.number;
+  }
+
+  toDOM() {
+    let container = document.createElement("span");
+
+    let doc = this.view.state.doc;
+    let pos = stepLocation(this.step);
+    let maxLineLen = 0;
+
+    for (let line of doc.iterLines()) {
+      maxLineLen = Math.max(maxLineLen, line.length);
+    }
+    let currLine = this.line;
+    let hideTable = !this.focusedLines.includes(currLine.number);
+    this.tableHidden = hideTable;
+
+    let padding = 2 + maxLineLen - currLine.length;
+    let spaces = " " + "_".repeat(padding);
+
+    let arrowOut = "»";
+    let arrowIn = "«";
+
+    let toggleHiddenPaths = (_e: any) => {
+      this.pathsHidden = !this.pathsHidden;
+      let content = container.querySelector<HTMLDivElement>(
+        ".step-table-content"
+      );
+      let dropdown = container.querySelector<HTMLDivElement>(
+        ".step-table-dropdown"
+      );
+      if (content) {
+        content.style.height = this.pathsHidden ? "0px" : "auto";
+      }
+      if (dropdown) {
+        dropdown.innerHTML = this.pathsHidden ? "..." : "";
+      }
+    };
+
+    let toggleWidget = (_e: any) => {
+      this.tableHidden = !this.tableHidden;
+      let contents = container.querySelector<HTMLDivElement>(
+        ".step-widget-contents"
+      );
+      let btn = container.querySelector<HTMLSpanElement>(
+        ".step-table-collapse"
+      );
+
+      if (contents) {
+        contents.style.width = this.tableHidden ? "0px" : "auto";
+      }
+      if (btn) {
+        btn.innerHTML = this.tableHidden ? arrowOut : arrowIn;
+      }
+    };
+    let matchAll = new RegExp("(.*)?");
+    let r = this.focusedPaths.get(currLine.number) ?? matchAll;
+    let [focusedDiffs, hiddenDiffs] = partition(
+      ([path, _]) => !!path.match(r),
+      this.step.state
+    );
+
+    let hiddenDropdown =
+      hiddenDiffs.length > 0 ? (
+        <>
+          <div className="step-table-dropdown">...</div>
+          <div className="step-table-content">
+            <table className="perm-step-table">
+              {hiddenDiffs.map(([path, diffs]) => (
+                <PermDiffRow path={path} diffs={diffs} />
+              ))}
+            </table>
+          </div>
+        </>
+      ) : null;
+
+    let rowStyle = {
+      width: hideTable ? "0px" : "auto",
+    };
+
+    ReactDOM.createRoot(container).render(
+      <div className="perm-step-widget">
+        <span className="step-table-collapse" onClick={toggleWidget}>
+          {hideTable ? arrowOut : arrowIn}
+        </span>
+        <div className="step-widget-contents" style={rowStyle}>
+          {spaces}
+          <div className="step-table-container" onClick={toggleHiddenPaths}>
+            <div className="step-table-parent">
+              <table className="perm-step-table">
+                {focusedDiffs.map(([path, diffs]) => (
+                  <PermDiffRow path={path} diffs={diffs} />
```

Comment:
In React, all components produced from an iterator need a `key`. The easiest thing is usually just to use the index, e.g. 

```tsx
.map(([path, diffs], i) => <PermDiffRow key={i} ... />
```